### ISSUE TITLE
Using dockerized wallabag installation behind a apache reverse proxy

### ISSUE THREAD

**Issue by MoriTanosuke** [wallabag/wallabag#2273]
### Issue details

I'm using the docker installation of Wallabag and another container with an Apache acting as a reverse proxy. I want my setup to be accessible from the outside as https://myhost/wallabag so I have this `Location` directive in my *apache2.conf*:

````
<Location "/wallabag/">
        Order Allow,Deny
        Allow from all
        ProxyPass                http://my_wallabag_container:80/
        ProxyPassReverse         http://my_wallabag_container:80/
</Location>
````

Both containers are in the same network, but only the apache container has exposed ports on my host.

When I try to open https://myhost/wallabag/ I'm redirected to http://myhost/login instead of https://myhost/wallabag/login. I can open https://myhost/wallabag/login and get the HTML of the login form, but all *href*s to images, CSS and JS are pointing to */bundles/...*

Example:
    
    <link rel="stylesheet" href="/bundles/wallabagcore/themes/material/css/materialize.min.css"  media="screen,projection"/>

The form action also points to the wrong URL:

    <form action="/login_check" method="post" name="loginform">

Should be 

    <form action="/wallabag/login_check" method="post" name="loginform">

This is basically the same behavior that I described in #1075 and provided in PR https://github.com/wallabag/wallabag/pull/1214

I don't know if it's already possible with Symfony to set the base URL of the installation to */wallabag* instead of */*.

**Comment by j0k3r** [comment]
You might to look into the vhost and/or the `web/.htaccess` file to see if you can defined a base url.

But it looks like to might be automatic in Symfony since v2.2: https://github.com/symfony/symfony-standard/pull/530 :confused:

**Comment by MoriTanosuke** [comment]
Will have a look into the `web/.htaccess`. Although https://github.com/symfony/symfony-standard/pull/497#issuecomment-16452221 looks promising.

**Comment by MoriTanosuke** [comment]
I fixed this be adding the following option to the `config.yml`:

````
framework:
    ...
    assets:
        base_urls: "%env.base_path%"
````

I then added the following option to my `docker-compose.yml`:

````
wallabag:
  image: wallabag/wallabag
  volumes:
  - ./wallabag/data:/var/www/wallabag/data
  - ./wallabag/config.yml:/var/www/wallabag/app/config/config.yml
  environment:
  - SYMFONY__ENV__BASE_PATH=https://myhost/wallabag/
````

Images, CSS and JS are now available.

But the login still does not work and redirects me to http://mycontainer/login instead of https://myhost/wallabag/login. Opening https://myhost/wallabag/login manually still gives me the login form, this time with correct styling.

Bit better, but still not optimal, because I can not login.

**Comment by MoriTanosuke** [comment]
I also tried to set the following

````
# app/config/parameters.yml
parameters:
    router.request_context.host: myhost
    router.request_context.scheme: https
    router.request_context.base_url: wallabag
````

But this didn't change anything. I'll try to override the settings in the `RequestContext` as mentioned on https://symfony.com/doc/current/console/request_context.html later.

**Comment by MoriTanosuke** [comment]
Maybe https://github.com/symfony/symfony/issues/19396 is related.

**Comment by MoriTanosuke** [comment]
Today I was idly scrolling through the *internal settings* of my installation, and I discovered that there is a setting to change the URL of my installation:

![wallabag_internal_settings_url](https://cloud.githubusercontent.com/assets/105346/18667223/f4000b28-7f2f-11e6-875e-64274e996d0a.png)

I modified this to match my URL, but unfortunately the situation is still the same. The images/css/js are displayed, but all links are still missing the base URL.

Can anybode explain to my what this internal setting is supposed to do?

**Comment by j0k3r** [comment]
This url is only used to image displaying (like the top logo, when you click on it the href use this parameter), also used in the activation email (so the link goes to your wallabag instance).
It does nothing more.

**Comment by callmetango** [comment]
@MoriTanosuke I had a "nice" weekend with the attempt to install Wallabag in a sublocation of my localhost. Here are my findings:

The redirection to /login is caused by the [Symfony Security Bundle](https://symfony.com/doc/current/reference/configuration/security.html). It has several [configuration options for the login form](https://symfony.com/doc/current/reference/configuration/security.html#the-login-form-and-process) like `login_path` and `check_path` which default to `/login` and `/login_check` resp.

So you have to change these paths as well as other paths of the Symfony firewall config and the routing to work with your `base_path`. See https://github.com/wallabag/wallabag/pull/2181 and especially https://github.com/wallabag/wallabag/pull/2181#issuecomment-279361238 for a solution.

I also suggest you change your current configuration of `app/config/config.yml` as given in https://github.com/wallabag/wallabag/issues/2273#issuecomment-245503869 like so:
```yaml
framework:
    # ...
    assets:
        base_urls: "%base_path%"
```
and then add the `base_path` parameter to the `app/config/parameters.yml` like so:
```yaml
parameters:
    base_path: "%env.base_path%"
    # ...
```
Though I haven't tested the parameter interpolation with a Docker setup this way you should have a consistent configurable base path in all your config files and you could keep using the solution once https://github.com/wallabag/wallabag/pull/2181 is merged. Please comment if this doesn't work.

**Comment by MoriTanosuke** [comment]
@callmetango nice, this sounds promising. I just switched my Installation over to the official docker image, but I will try your findings and see if it solves my issue

I'm still running my installation with an exposed port without a subdirectory, so I'm not happy with my setup yet. Maybe your changes are the missing link. 👍

**Comment by rouja** [comment]
Hi,

I just created a wallabag container with apache to serve Wallabag. Then I started my container on my server and expose the service with apache configured as a reverse proxy in https. I got somes bad redirections between https and http.

For instance, after the login page, I was redirected on **http:**//wallabag.example.com instead of **https://**. It was the same behaviour during the logout.

To fix the issue, I found this documentation http://symfony.com/doc/current/request/load_balancer_reverse_proxy.html. 

So I juste added the IP of my reverse proxy like this in **app/config/config.yml** : 

`trusted_proxies: [172.17.0.1]`

And now everything seems good.

FYI I paste here the apache configuration (for the reverse proxy) : 

```
<VirtualHost wallabag.example.com:443>
  ServerName wallabag.example.com
  SSLEngine on
  SSLCertificateFile /etc/letsencrypt/live/wallabag.example.com/fullchain.pem
  SSLCertificateKeyFile /etc/letsencrypt/live/wallabag.example.com/privkey.pem

  ProxyPreserveHost On
  ProxyPass / http://172.17.0.1:10000/
  ProxyPassReverse / http://172.17.0.1:10000/


  ErrorLog ${APACHE_LOG_DIR}/wallabag-error.log
  CustomLog ${APACHE_LOG_DIR}/wallabag-access.log combined
</VirtualHost>
```
Maybe it would be nice to have a dedicated section for reverse-proxy in the documentation http://doc.wallabag.org/en/master/ ?

Bye

**Comment by towo** [comment]
Cheers @rouja,

an important aside: cache needs to be cleared with `make clean` afterwards, or you'll spend hours debugging why the hell it doesn't work.

**Comment by crbble** [comment]
"make" is not available in the docker image, but you can remove the _/var/www/wallabag/var/cache_ inside the image.

Nevertheless, I also set up a dockerized image with nginx as reverse proxy and in my configuration it was enough to set the slightly understressed SYMFONY__ENV__DOMAIN_NAME= variable to a full URL including https:// and a different port (in my case).

**Comment by stenwt** [comment]
> This url is only used to image displaying (like the top logo, when you click on it the href use this parameter), also used in the activation email (so the link goes to your wallabag instance).
It does nothing more.

It seems really odd to me to use SYMFONY__ENV__DOMAIN_NAME to fully-qualify the URLs for images, etc. but then to hard-code app-links like /register and /login. I'm hitting the same problem with the current docker image, where styles/images/etc load fine, but I can't login or register because the links are broken.

**Comment by TomFaulkner** [comment]
It looks like I'm having the same issue. https://github.com/wallabag/docker/issues/104

Doing similar, using Caddy in Docker.

**Comment by b-m-f** [comment]
What fixed it for me with caddy and docker was setting all of the following headers in the proxy definition in caddy:
```
    header_upstream Host {host}
    header_upstream X-Forwarded-Host {host}
    header_upstream X-Real-IP {remote}
    header_upstream X-Forwarded-For {remote}
    header_upstream X-Forwarded-Proto {scheme}
```
Im sure this should work for other reverse proxies as well.

**Comment by TomFaulkner** [comment]
@b-m-f Thanks, I'll give it a try.

**Comment by TomFaulkner** [comment]
I tried adding that to Caddy. I think that helps, as I am redirected to the proper URL now.

Unfortunately, now I have a different issue that seems to be a database issue. I think I'll try moving to Redis rather than Postgres later.

Thanks.

**Comment by luckydonald** [comment]
Using the nginx [proxy_redirect](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_redirect) similar to #3260, together with the [comment above](https://github.com/wallabag/wallabag/issues/2273#issuecomment-372149830) worked:
```nginx
listen 443 ssl;
server_name wallabag.example.com;
set $upstream http://wallabag;  # container name

location / {
        proxy_set_header X-Forwarded-Host $server_name;
        proxy_set_header X-Forwarded-Proto https;
        proxy_set_header X-Forwarded-For $remote_addr;
        proxy_redirect $upstream https://$server_name;
        # proxy_redirect http://wallabag https://wallabag.example.com
}
```

**Comment by diegombeltran** [comment]
> Hi,
> 
> I just created a wallabag container with apache to serve Wallabag. Then I started my container on my server and expose the service with apache configured as a reverse proxy in https. I got somes bad redirections between https and http.
> 
> For instance, after the login page, I was redirected on **http:**//wallabag.example.com instead of **https://**. It was the same behaviour during the logout.
> 
> To fix the issue, I found this documentation http://symfony.com/doc/current/request/load_balancer_reverse_proxy.html.
> 
> So I juste added the IP of my reverse proxy like this in **app/config/config.yml** :
> 
> `trusted_proxies: [172.17.0.1]`
> 
> And now everything seems good.
> 
> FYI I paste here the apache configuration (for the reverse proxy) :
> 
> ```
> <VirtualHost wallabag.example.com:443>
>   ServerName wallabag.example.com
>   SSLEngine on
>   SSLCertificateFile /etc/letsencrypt/live/wallabag.example.com/fullchain.pem
>   SSLCertificateKeyFile /etc/letsencrypt/live/wallabag.example.com/privkey.pem
> 
>   ProxyPreserveHost On
>   ProxyPass / http://172.17.0.1:10000/
>   ProxyPassReverse / http://172.17.0.1:10000/
>   RequestHeader set X-Forwarded-Proto "https"
> 
> 
>   ErrorLog ${APACHE_LOG_DIR}/wallabag-error.log
>   CustomLog ${APACHE_LOG_DIR}/wallabag-access.log combined
> </VirtualHost>
> ```
> 
> Maybe it would be nice to have a dedicated section for reverse-proxy in the documentation http://doc.wallabag.org/en/master/ ?
> 
> Bye

This fixed the CLEARTEXT error while trying to connect with the Android app. I thought my reverse proxy (Traefik) was not working properly at first, but then saw this:

```
[~]$ curl -I https://domain.com                                                                                                                                                                                                                                                                           
HTTP/2 302                                                                                                                                                                                                                                                                                                                     
cache-control: no-cache, private                                                                                                                                                                                                                                                                                               
content-type: text/html; charset=UTF-8                                                                                                                                                                                                                                                                                         
date: Mon, 21 Jan 2019 17:42:31 GMT                                                                                                                                                                                                                                                                                            
location: http://domain.com/login
```

So I "DockerExec ed" to the Wallabag container, modified the file and added the trusted_proxies line with the Traefik container internal IP, restarted it, and voilá.

**Comment by sztanpet** [comment]
> > So I juste added the IP of my reverse proxy like this in **app/config/config.yml** :
> > `trusted_proxies: [172.17.0.1]`

Where does that option go in the file?

**Comment by sztanpet** [comment]
ah, found it, under `framework`! thanks!

**Comment by norweeg** [comment]
this option is apparently deprecated (see https://github.com/wallabag/wallabag/commit/80336f77fd6ef72dafe20b35b74d74d06592c06f ) .  We need a better fix than this.

**Comment by j0k3r** [comment]
Regarding the documentation, you can still use trusted_proxies, but you need to update `web/app.php` file and add this line (and update first argument values) just after the `$kernel = new AppKernel('prod', false);`:

```php
Request::setTrustedProxies(['192.0.0.1', '10.0.0.0/8'], Request::HEADER_X_FORWARDED_ALL);
```

It this can fix the problem, we might introduce a new parameter in `app/config/parameters.yml` so people can define their proxies without editing the code.

**Comment by norweeg** [comment]
for what it's worth, it is possible to fix this using your proxy without changing anything, but yeah, it would be nice to be able to configure it

**Comment by j0k3r** [comment]
@norweeg can you give it a try and tell me? Then I'll push a fix for the 2.4.0

**Comment by norweeg** [comment]
@j0k3r  which method did you want me to try?  I can confirm that setting the trusted_proxies: key of the framework section of parameters.yml *does* work per the other issue comment threads, but even with the ability to  set this variable via an environment variable is tricky.  If you're bringing up an instance of wallabag with a reverse proxy via `docker-compose up -d`  like I do, you may not know what ip the reverse proxy will have until it is created and I can't get it's ip set somewhere that docker-compose can find it when bringing up the wallabag container.  While I am able to work-around the issue by making my proxy handle the redirects rather than relying on the wallabag container, it would be better if the wallabag container didn't downgrade from https to http when redirecting to /login

**Comment by devzsolt** [comment]
> What fixed it for me with caddy and docker was setting all of the following headers in the proxy definition in caddy:
> 
>     header_upstream Host {host}
>     header_upstream X-Forwarded-Host {host}
>     header_upstream X-Real-IP {remote}
>     header_upstream X-Forwarded-For {remote}
>     header_upstream X-Forwarded-Proto {scheme}
> 
> Im sure this should work for other reverse proxies as well.


If someone is using caddy-docker-proxy, then this should be fine without additional hacks, it does the same as described above by @b-m-f :

```
    labels:
      caddy.address: "yourdomain.com"
      caddy.targetport: "80"
      caddy.proxy.transparent: ""

```

**Comment by joonhshin** [comment]
> > What fixed it for me with caddy and docker was setting all of the following headers in the proxy definition in caddy:
> > ```
> > header_upstream Host {host}
> > header_upstream X-Forwarded-Host {host}
> > header_upstream X-Real-IP {remote}
> > header_upstream X-Forwarded-For {remote}
> > header_upstream X-Forwarded-Proto {scheme}
> > ```
> > 
> > 
> > Im sure this should work for other reverse proxies as well.
> 
> If someone is using caddy-docker-proxy, then this should be fine without additional hacks, it does the same as described above by @b-m-f :
> 
> ```
>     labels:
>       caddy.address: "yourdomain.com"
>       caddy.targetport: "80"
>       caddy.proxy.transparent: ""
> ```

Can you provide examples for the {host}, {remote}, {scheme} parameters? For example, is {host} the docker host public IP address? And what goes in {remote} and {scheme}? Thanks!

**Comment by cmuench** [comment]
If someone uses Traefik 2.0 with Letsencrypt to terminate SSL as reverse proxy.

Add this labels to your wallabag container config:

```
labels:
  - "traefik.http.routers.wallabag.middlewares=wallabagHeaders"
  - "traefik.http.middlewares.wallabagHeaders.headers.sslProxyHeaders.X-Forwarded-Proto=https"
```

**Comment by Jenjen1324** [comment]
> 
> 
> @j0k3r which method did you want me to try? I can confirm that setting the trusted_proxies: key of the framework section of parameters.yml _does_ work per the other issue comment threads, but even with the ability to set this variable via an environment variable is tricky. If you're bringing up an instance of wallabag with a reverse proxy via `docker-compose up -d` like I do, you may not know what ip the reverse proxy will have until it is created and I can't get it's ip set somewhere that docker-compose can find it when bringing up the wallabag container. While I am able to work-around the issue by making my proxy handle the redirects rather than relying on the wallabag container, it would be better if the wallabag container didn't downgrade from https to http when redirecting to /login

From the symfony docs, can't we just use REMOTE_ADDR as the proxy IP parameter? Though we should add to the docs that when you use that value you have to make sure it's not accessible without the reverse proxy.

https://symfony.com/doc/master/deployment/proxies.html#but-what-if-the-ip-of-my-reverse-proxy-changes-constantly

**Comment by norweeg** [comment]
@Jenjen1324 yeah, that should work

**Comment by gerroon** [comment]
HI

I am on Apache and I have a working docker wallabag. Whis solution is valid as of now? A lot of ideas were recommended with half working solutions and some are for Nginx and such. 

So far all I get is redirecting to login page without wallabag path :(

**Comment by Jenjen1324** [comment]
> 
> 
> HI
> 
> I am on Apache and I have a working docker wallabag. Whis solution is valid as of now? A lot of ideas were recommended with half working solutions and some are for Nginx and such.
> 
> So far all I get is redirecting to login page without wallabag path :(

My working install with the `wallabag/wallabag:2.3.8` image (current latest):

Docker image environment variables:

    SYMFONY__ENV__DOMAIN_NAME=https://my.public.host.name

I use nginx as my reverse proxy, but I didn't need to do anything different from any other reverse proxy setup I had. But FWIW, I configured the following headers (as I do for all my proxied sites): `Host, Referer, Origin, X-Real-IP, X-Forwarded-Host, X-Forwarded-Server, X-Forwarded-Port, X-Forwarded-For, X-Forwarded-Proto, X-Forwarded-Protocol`  
I also set up the http upgrade stuff for websockets, but I'm not sure if that's relevant for wallabag.

Here's my full nginx config https://gist.github.com/Jenjen1324/860b4a3486de063f422b1dfe61cf7c05

I'm not sure if this line is required: `proxy_redirect "http://wallabag.redacted.tld/" "https://wallabag.redacted.tld/";` but I think it was to just rewrite all http urls to https or something.

**Comment by gerroon** [comment]
@Jenjen1324 

Thanks for the reply but it seems like you are using a subdomain, my issue is with using it behind a sub path like `https://DOMAIN/wallabag`

**Comment by Sparticuz** [comment]
No Subdirectory/subpath is my issue too. It's because the paths are hard coded in wallabag and not configurable. Both the PRs addressing this have been closed. https://github.com/wallabag/wallabag/pull/2895 and https://github.com/wallabag/wallabag/pull/2181

**Comment by jt196** [comment]
FYI just setting Wallabag up and had the same issue with Android login. For those with nginx server setups, the line that worked for me was the redirect in @Jenjen1324's [set up](https://gist.github.com/Jenjen1324/860b4a3486de063f422b1dfe61cf7c05):

`proxy_redirect "http://wallabag.redacted.tld/" "https://wallabag.redacted.tld/";`

**Comment by Tukks** [comment]
> 
> 
> No Subdirectory/subpath is my issue too. It's because the paths are hard coded in wallabag and not configurable. Both the PRs addressing this have been closed. #2895 and #2181



> 
> 
> @Jenjen1324
> 
> Thanks for the reply but it seems like you are using a subdomain, my issue is with using it behind a sub path like `https://DOMAIN/wallabag`

Same problem here with subpath, do you have found any solution with nginx?

**Comment by ngosang** [comment]
Do not lose your time. It can't be fixed with Apache/Nginx configurations.
Several changes in the source code are required. This was a nice try but is not working => #2895
My fix was to setup a sub domain and keep the base_path in /.

**Comment by mvaldes14** [comment]
> 
> 
> If someone uses Traefik 2.0 with Letsencrypt to terminate SSL as reverse proxy.
> 
> Add this labels to your wallabag container config:
> 
> ```
> labels:
>   - "traefik.http.routers.wallabag.middlewares=wallabagHeaders"
>   - "traefik.http.middlewares.wallabagHeaders.headers.sslProxyHeaders.X-Forwarded-Proto=https"
> ```
> 
> (Replace "routers.**wallabag**" with your router name)

Added this setup to mine and still no dice, even got some redirects. Still broken. Kind of regret making it https now.

`      
        - "traefik.http.middlewares.wallabagHeaders.headers.sslProxyHeaders.X-Forwarded-Proto=https"
        - "traefik.http.middlewares.wb-redirect.redirectregex.regex=^http://mydomain.com/"
        - "traefik.http.middlewares.wb-redirect.redirectregex.replacement=https://mydomain.com"
        - "traefik.http.routers.wallabag.middlewares=wallabagHeaders@docker,wb-redirect@docker"
`

**Comment by fangfufu** [comment]
This is what I have for my Apache, Apache basically forces URL redirection. 
```
<VirtualHost _default_:80>
ServerName wallabag.YOUR.DOMAIN

ServerSignature Off

RewriteEngine On
RewriteRule ^ https://%{SERVER_NAME}%{REQUEST_URI} [END,NE,R=permanent]

ErrorLog /var/log/apache2/redirect.error.log
LogLevel warn
</VirtualHost>
<IfModule mod_ssl.c>
        <VirtualHost _default_:443>
                ServerName wallabag.YOUR.DOMAIN
                ProxyPreserveHost On
                ProxyPass / http://127.0.0.1:12345/
                ProxyPassReverse / http://127.0.0.1:12345/
                Include /etc/letsencrypt/options-ssl-apache.conf
                RequestHeader set X-Forwarded-Proto https
                ProxyPreserveHost On
                SSLCertificateFile /etc/letsencrypt/live/YOUR.DOMAIN/fullchain.pem
                SSLCertificateKeyFile /etc/letsencrypt/live/YOUR.DOMAIN/privkey.pem
        </VirtualHost>
</IfModule>
```

It works fine on the browser.

**Comment by cf8865042** [comment]
I had been strugling with this issue for days.  Finally I found a very rude and straigthforward solution that can get my dockerized wallabag served with prefix path behind Apache2 reverse proxy. This would be a solution if you are playing with Wallabag yourself. Hope it can helps.

It would be a solution if you satisfy following situation:
1. You want to serve your Wallabag with container.( docker pull wallabag/wallabag:latest ,  As of now 2.5.2 )
2. Apache server is serving your apps as reverse proxy server on Host machine.
3. You want to serve your  Wallabag with prefix path. (e.g. https://your_domain.com/wallabag/)

Solution first:
1. Symfony file 'Symfony/Component/HttpFoundation/Request.php' is taking care of generating the request paths. So we can tweak this file to add prefix for request paths.
   Remark: as per the article [Webcache Poisoning via X-Forwarded-Prefix and sub-request](https://symfony.com/blog/cve-2021-41267-webcache-poisoning-via-x-forwarded-prefix-and-sub-request), X-Forwarded-Prefix was added from Symfony 5.2. But the version of Symfony from the Wallabag image is 3.4. In other words, the underlying Symfony from Wallabag does not support reverse proxy with prefix. For reference , see the latest  [Request class](https://github.com/symfony/symfony/blob/acb7a989976f5f88cf745e9580a4ece23d65e12a/src/Symfony/Component/HttpFoundation/Request.php) from Symfony.
2. Reference a parameter value in Request.php to avoid hardcoding the application prefix. Define the parameter value in nginx location block.

Implement the solution as the following:
1.Bring up the Wallabag container(Generally the docker-compose). Copy following files from inside the Wallabag container :
`/var/www/wallabag/vendor/symfony/symfony/src/Symfony/Component/HttpFoundation/Request.php
/etc/nginx/nginx.conf`
to host OS path:
e.g.  
`/opt/wallabag/vendor/symfony/symfony/src/Symfony/Component/HttpFoundation/Request.php
/opt/wallabag/nginx/nginx.conf`

2. Stop the Wallabag container.

3. Edit host OS file "/opt/wallabag/vendor/symfony/symfony/src/Symfony/Component/HttpFoundation/Request.php" . 
  a. Locate the method 'getBaseUrl()' and change it to:
  `public function getBaseUrl()
    {
        //if (null === $this->baseUrl) {
        //    $this->baseUrl = $this->prepareBaseUrl();
        //}

        if (\array_key_exists('ZIEG_BASE_PATH', $_SERVER)) {
            return $this->getBaseUrlReal().$_SERVER['ZIEG_BASE_PATH'];
        }

        return $this->getBaseUrlReal();
    }`

  b. Add a new method 'getBaseUrlReal()' right below the method 'getBaseUrl()':
  `private function getBaseUrlReal()
    {
        if (null === $this->baseUrl) {
            $this->baseUrl = $this->prepareBaseUrl();
        }

        return $this->baseUrl;
    }`
   
  c. Locate the method 'preparePathInfo()' and substitutes the reference  '$this->baseUrl' to ' $this->getBaseUrlReal()':
  ` protected function preparePathInfo()
    {
        if (null === ($requestUri = $this->getRequestUri())) {
            return '/';
        }

        // Remove the query string from REQUEST_URI
        if (false !== $pos = strpos($requestUri, '?')) {
            $requestUri = substr($requestUri, 0, $pos);
        }
        if ('' !== $requestUri && '/' !== $requestUri[0]) {
            $requestUri = '/'.$requestUri;
        }

        if (null === ($baseUrl = **$this->getBaseUrlReal()**)) {
            return $requestUri;
        }

        $pathInfo = substr($requestUri, \strlen($baseUrl));
        if (false === $pathInfo || '' === $pathInfo) {
            // If substr() returns false then PATH_INFO is set to an empty string
            return '/';
        }

        return (string) $pathInfo;
    }`

4. Edit Host OS file '/opt/wallabag/nginx/nginx.conf':
  a. Add 'fastcgi_param ZIEG_BASE_PATH "/your_wallabag_prefix";' within the location block.
  b. Add 'fastcgi_param  HTTPS 'on';' if you serve Wallabag over SSL. 

5. Edit docker-compose.yml add above two files mapping under volumes of Wallabag service:
 `volumes:
      - /opt/wallabag/images:/var/www/wallabag/web/assets/images
      - /opt/wallabag/nginx/nginx.conf:/etc/nginx/nginx.conf
      - /opt/wallabag/vendor/symfony/symfony/src/Symfony/Component/HttpFoundation/Request.php:/var/www/wallabag/vendor/symfony/symfony/src/Symfony/Component/HttpFoundation/Request.php`

6. Add reverse proxy to Apache virtual host:
        `ProxyPass /wallabag/ http://localhost:***/
        ProxyPassReverse /wallabag/ http://localhost:***/`

7. Bring up the Wallabag container then your Wallabag would be served with prefix behind apache2.

**Comment by 7heo** [comment]
So there are two different groups of people commenting on this issue:

- People who want to install wallabag in `$scheme://their.server/wallabag-or-some-other-path/`
- People who want to serve wallabag over `HTTPS`.

For the latter, the problem seems solved. For the former, however, it still remains. Read on.

>@MoriTanosuke:
> I don't know if it's already possible with Symfony to set the base URL of the installation to /wallabag instead of /.

AFAIK it is, just not implemented in wallabag.

>@MoriTanosuke:
> Today I was idly scrolling through the _internal settings_ of my installation, and I discovered that there is a setting to change the URL of my installation:
> 
> 8<--------- SNIP --------->8
> 
> I modified this to match my URL, but unfortunately the situation is still the same. The images/css/js are displayed, but all links are still missing the base URL.
> 
> Can someone explain to me what this internal setting is supposed to do?

>@j0k3r:
> This url is only used to image displaying (like the top logo, when you click on it the href use this parameter), also used in the activation email (so the link goes to your wallabag instance). It does nothing more.

That's unfortunate...

>@callmetango:
> I had a "nice" weekend with the attempt to install Wallabag in a sublocation of my localhost. Here are my findings:
> 
> The redirection to /login is caused by the [Symfony Security Bundle](https://symfony.com/doc/current/reference/configuration/security.html). It has several [configuration options for the login form](https://symfony.com/doc/current/reference/configuration/security.html#the-login-form-and-process) like `login_path` and `check_path` which default to `/login` and `/login_check` resp.
> 
> So you have to change these paths as well as other paths of the Symfony firewall config and the routing to work with your `base_path`. See #2181 and especially [#2181 (comment)](https://github.com/wallabag/wallabag/pull/2181#issuecomment-279361238) for a solution.
> 
> I also suggest you change your current configuration of `app/config/config.yml` as given in [#2273 (comment)](https://github.com/wallabag/wallabag/issues/2273#issuecomment-245503869) like so:
> 
> 8<--------- SNIP --------->8

Once the httpd is properly configured, I'm sure the remaining problems have to do with the routing, and potentially the security, yes.

>@Sparticuz:
> No Subdirectory/subpath is my issue too. It's because the paths are hard coded in wallabag and not configurable. Both the PRs addressing this have been closed. #2895 and #2181

That was my experience too, yes.

>@Tukks:
> @Sparticuz @gerroon
> 
> > No Subdirectory/subpath is my issue too. It's because the paths are hard coded in wallabag and not configurable. Both the PRs addressing this have been closed. #2895 and #2181
> 
> > @Jenjen1324
> > Thanks for the reply but it seems like you are using a subdomain, my issue is with using it behind a sub path like `https://DOMAIN/wallabag`
> 
> Same problem here with subpath, do you have found any solution with nginx?

>@ngosang:
> Do not lose your time. It can't be fixed with Apache/Nginx configurations. Several changes in the source code are required. This was a nice try but is not working => #2895 My fix was to setup a sub domain and keep the base_path in /

I have found a working solution with nginx (see below), but it is hacky. At the same time, it works.

>@cf8865042:
> I had been strugling with this issue for days. Finally I found a very rude and straigthforward solution that can get my dockerized wallabag served with prefix path behind Apache2 reverse proxy. This would be a solution if you are playing with Wallabag yourself. Hope it can helps.
>
> It would be a solution if you satisfy following situation:
>
> 1. You want to serve your Wallabag with container.( docker pull wallabag/wallabag:latest , As of now 2.5.2 )
> 2. Apache server is serving your apps as reverse proxy server on Host machine.
> 3. You want to serve your Wallabag with prefix path. (e.g. https://your_domain.com/wallabag/)

Kudos for trying, but your solution is pretty complex, and requires tempering with the container (therefore making it hard to maintain).

---

Here is my solution:

```nginx
        # In nginx.conf or any nginx conf file, use the following snippet,
        # eventually replacing 127.0.0.1 by the IP address that serves the
        # wallabag app, and <your_wallabag_port> with its TCP port.
        # The snippet is made to work for any <sub_directory>, in my case I have
        # tried with "wallabag", making the location rule:
        #     location /wallabag/ {

        # The trailing / is essential for routing to work.
        location /<sub_directory>/ {
                # Same here, the trailing / is essential for the rewrite to
                # properly work. Don't forget to change http to https if your
                # backend uses HTTPS.
                proxy_pass http://127.0.0.1:<your_wallabag_port>/;
                # The proxy_redirect instructions are necessary to "catch and
                # rewrite" the "Location:" headers returned by the app. We have
                # them once for each protocol, in order to catch rewrites to
                # any protocol, and rewrite it to the current one.
                proxy_redirect http://$server_name/ $scheme://$server_name/<sub_directory>/;
                proxy_redirect http://$server_name/login $scheme://$server_name/<sub_directory>/login;
                proxy_redirect https://$server_name/ $scheme://$server_name/<sub_directory>/;
                proxy_redirect https://$server_name/login $scheme://$server_name/<sub_directory>/login;
                # The following is a hack to tell the proxy not to compress
                # anything, so to be able to rewrite the responses on the fly.
                proxy_set_header Accept-Encoding "";
                # The sub_filter instructions replace patterns in the returned
                # response. This is the actual hack that makes this entire
                # solution work.
                sub_filter 'action="/' 'action="/<sub_directory>/';
                sub_filter 'href="/' 'href="/<sub_directory>/';
                sub_filter 'src="/' 'src="/<sub_directory>/';
                # We also rewrite any link to the current server to use the
                # currently used protocol.
                sub_filter 'https://$server_name' '$scheme://$server_name';
                sub_filter 'http://$server_name' '$scheme://$server_name';
                # We replace matches as many times as we can.
                sub_filter_once off;
                # This is necessary so that wallabag does not return links to
                # the backend.
                proxy_set_header Host $host;
                # Standard proxy instructions.
                proxy_http_version 1.1;
                proxy_set_header X-Forwarded-Host $host;
                proxy_set_header X-Forwarded-Proto $scheme;
        }
```

**Comment by 7heo** [comment]
@gerroon if you already copy/pasted that config into your nginx, please update it, I have added an additional rule for the /annotations XHR requests, so that they properly work.

**Comment by gerroon** [comment]
My issue is that I am now on Apache :(
